### DISCUSSION TITLE
A Complex Algorithm - how to with dask and/or spark

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#197] [Design & Development]
I decided to start a discussion page on this topic because I think it has some elements that would be educational to future users and may present some challenges that could impact our design.   The "challenges" may just be my ignorance so bear with me.

For the record, what I'm working on is conversion of one the element so the plane wave migration code called pwstack.  The old pwstack program was implemented in C++ and went through a couple of stages of development that turned it into a monster.  The redesign, however, reduces the algorithm to these stages:

1. Query the source collection for a set of sources defined by an earlier stacking step (source side stack).  Note a variant would I have not tried is possible but beside the point for now.  The output of this query is a cursor that  could converted to a list of object ids in a trivial amount of memory.   
2. Each source_id maps to an expanded table with a set of tuples with the following content:  source_id, ix, iy, lat0, lon0.   ix-iy and lat0-lon0 are somewhat redundant but necessary to carry along.  ix,iy define a range ix=1,nx and iy=1,ny (rectangle) and lat0,lon0 are geographical points associated with each grid index (i.e. lat0(ix,iy) and lon0(ix,iy)).   Note  the expanded table is one (source_id) to many tuples (nx*ny tuples per source_id with the tuples duplicated exactly for each source_id).  
3.  The next step is a query to MongoDB for each row in the table above. The query is a match for source_id and
 a geospatial 2dsphere search around lat0,lon0 of waveforms linked to stations close to (defined by the circle/2dsphere) lat0,lon0.The query would probably best be cast as a map from each row of the tab to an output of MongoDB cursors.  
4. The working dataset would be created by calling the ensemble reader on each  cursor.   Alternatively, maybe this could be combined with step 3.  Beyond my understanding of dask and spark at this stage.
5.  A C++ function extracted from the old code would then be called on each ensemble - a map taking ensembles in and ensembles out.  

I'm still very early in my understanding of dask and/or spark so bear with me.  I have these questions I need help with:
1.  I cannot find any example on line that show how to do step 2 above?  i.e. how is a bag/rdd expanded with a one to many replationship?  That is the inverse of reduce. 
2. In step 3 is it possible to have a bag/rdd made up of a set of MongoDB cursors?
3. We may want to add some basic functions for doing some geospatial queries with MongoDB.   As I develop this I think it will be natural to learn what they are.  This is a side issue for now but one worth putting into the record.
4. I think step 4 requires a variant of read_distributed_data where the granularity is by ensemble.   This is probably not a unique example where this would be useful - any time the input is naturally grouped into an ensembles it would make sense.  Perhaps read_distributed_data could be extended to switch to calls to read_ensemble_data instead of read_data (which is what it does now) when it detects this kind of input.  
5. In reading the dask documentation it strikes me that the first 3 steps would map more naturally into dask dataframes instead of "bags".  In the online documentation I found a reference for a [bag to data frame converter](https://docs.dask.org/en/latest/bag-creation.html) but not the inverse (required I think to mesh with mspass).   Maybe this is a searching error as it strikes me from what I read that dataframe to bag is the easy conversion.  Can you educate me?  Does spark have a dataframe?

**Comment by wangyinz** [comment]
I think these are all very good example questions that our future users may ask. I am not sure that I have an exact answer for all of them, but this discussion can surely help us to shape the api. Anyway, see below:

> 1. I cannot find any example on line that show how to do step 2 above?  i.e. how is a bag/rdd expanded with a one to many replationship?  That is the inverse of reduce.

You should think of bag or rdd as a (distributed) list that doesn't guarantee ordering. The element of the list can be anything and is the granularity of parallelism. So, the answer to this question will be, we just need to put the "source_id : expanded_table" into a proper data structure, and arrange that into the list (bag/rdd). e.g., `[source_id1 : expanded_table1, source_id2 : expanded_table2, ...]`. This has nothing to do with reduce. To get step 2 from step 1, we are basically doing a map that changes the content of the list from `[source_id1, source_id2, ...]` to `[source_id1 : expanded_table1, source_id2 : expanded_table2, ...]`.

> 2. In step 3 is it possible to have a bag/rdd made up of a set of MongoDB cursors?

Just like above. Yes. 

> 3. We may want to add some basic functions for doing some geospatial queries with MongoDB.   As I develop this I think it will be natural to learn what they are.  This is a side issue for now but one worth putting into the record.

Yes, I agree.

> 4. I think step 4 requires a variant of read_distributed_data where the granularity is by ensemble.   This is probably not a unique example where this would be useful - any time the input is naturally grouped into an ensembles it would make sense.  Perhaps read_distributed_data could be extended to switch to calls to read_ensemble_data instead of read_data (which is what it does now) when it detects this kind of input.

Because the atomic level of parallelism is already the cursor in this step, we can simply call the ensemble reader here to map that bag of cursor into a bag of ensembles. There is no new methods needed here. The `read_distributed_data` method is for a very different scenario where you want to start a parallel workflow from the read. Here, everything is already in parallel, so we just call the sequential function and it will be run in parallel by the mappers.

> 5. In reading the dask documentation it strikes me that the first 3 steps would map more naturally into dask dataframes instead of "bags".  In the online documentation I found a reference for a [bag to data frame converter](https://docs.dask.org/en/latest/bag-creation.html) but not the inverse (required I think to mesh with mspass).   Maybe this is a searching error as it strikes me from what I read that dataframe to bag is the easy conversion.  Can you educate me?  Does spark have a dataframe?

Dataframe is more like a table, so it is indeed a better fit for the structure of data you described here. We didn't implement our api with the dataframes only because Spark's dataframe does not support customized data type, so we cannot put our data object in it. In fact, my original design was for dataframes. Anyway, for Dask, it doesn't really matter whether our api is built for bags or dataframes as converting between them is easy (i.e., [dask.bag.Bag.to_dataframe](https://docs.dask.org/en/latest/bag-api.html#dask.bag.Bag.to_dataframe) and [dask.dataframe.DataFrame.to_bag](https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.to_bag)). Still, using either container for the parallel workflow can achieve the goal here because the atomic data objects are equivalent anyway.

**Comment by pavlis** [comment]
Very helpful, thank you.   I recall the thing about dataframes in the past but did not remember that somewhat subtle issue.   Now I understand deeply why dask is the default.   An important point about the container, I think, is that for a higher level processing model like pwstack and pwmig it should be more than acceptable to have them only run with dask.  I think the dataframe capability make that the best choice for this algorithm.  Is there any reason not to think that?

I also realized another parallel construct that seems to me would make big difference in performance here that should be very common for any ensemble processing.  That is, a sensible parallel list would be this:  [query1, query2, ..., queryn] where the elements of the list are queries for an individual ensemble.   That would naturally parallelized the database query operation.   Would that be prudent?

On geospatial queries - a couple functions are an obvious need.   Here are some prototypes:
```
def setgeo(collection,lat_key='lat',lon_key='lon',geo_key='coords',build_index=True):
def geosearch(collection,domain,geo_key='coords):
```
the first is needed to take name-value pairs and produce the required struct for the geographic search find algorithms.  As described [here](https://docs.mongodb.com/manual/geospatial-queries/) geo searchs prefer a dict with a key (like coords) and an array of 2 number for x and y.   Our schema doesn't currently define those.  Since defining geo points like that means one will likely want to use it for searches the default should automatically create an index.     That function would return a cursor.

The second function proposed (geosearch) would be another example of exploiting obspy existing capabilities.  They have this [generic geographic region object ](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.mass_downloader.domain.Domain.html#obspy.clients.fdsn.mass_downloader.domain.Domain) that appears to provide a container that can be directly translated into geospatial queries.  

Normalized collections are going to be a complication here.   Our design has largely put all coordinate data in the (smaller) site, channel, and source collections.   If geospatial queries were the norm it might be smart to provide a function that would not write the array (coords  default key above) into the normalizing collection but in the wf collection after doing a cross-reference.   Then wf collection searchs, I would assume, would be much faster.  Otherwise constructing a geospatial query into a wf collection gets a bit ugly.   e.g. to get data with a set of sources in a specific region and a set of receivers in another region you would have to do a spatial query on both source and receiver collections and then some complicated query using a list of source_ids and site_id/channel_id fields.   I can't imagine that not be a ponderously slow query.   Thus, maybe setgeo should have this signature that is a variant of above:
```
def setgeo(wf_collection='wf_TimeSeries', normalization_collection='channel',lat_key='lat',lon_key='lon',geo_key='chan_coords',build_index=True):
```
where this would upsert (by default) a chan_coords array pair in each wf_TimeSeries document.  

Opinions?

**Reply by pavlis** [reply]
PS:   [here](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.mass_downloader.domain.html#module-obspy.clients.fdsn.mass_downloader.domain) is a page showing implementations in obspy of the abstract base Domain exist.

**Comment by wangyinz** [comment]
Yes, it is fine for external algorithms to support only one of the parallel api. Actually, another decision we might eventually make is whether it worth it to support both Spark and Dask. I guess we need to run some benchmarks first.

For the parallel query, right now, our implementation already supports that. If the element of a bag/rdd is a query, we can directly call our ensemble reader with a map call to map those queries into ensembles. I guess what's missing is a way to construct such queries. I guess we could generalize this concept and make a query builder that returns a list of queries (by queries I mean id list that can pass to the ensemble reader). I don't know exactly what that looks like, but you probably have an example already from pwstack.

The api for geospatial queries looks good to me. We should probably use the GeoJSON objects instead of the legacy coordinate pairs. The ObsPy's Domain object looks pretty handy, and we should definitely build on that. While I agree adding the `chan_coords` to wf collection is probably the best approach, I am not sure that a normalized solution will be that slow. I think (which may be wrong) as long as we index the source_id and site_id/channel_id fields in the wf collection, the query should perform OK. No matter what, we definitely should implement api you posted here. Let's move it into a design doc.

**Reply by pavlis** [reply]
Looking at this a bit more this morning with these two followups:
1.  Good point about GeoJSON objects - I had misread the documentation in the all to common quick scan.
2.  I quickly implemented a way to translate an obspy [CircularDomeain](https://docs.obspy.org/packages/autogen/obspy.clients.fdsn.mass_downloader.domain.CircularDomain.html#obspy.clients.fdsn.mass_downloader.domain.CircularDomain) to a MongoDB geospatial query.  The other two methods available are RectangularDomain and the base class Domain.   I'm not as sure how to do the later two and will punt this for now.   Let's treat what I'm writing now as a prototype we can use to build a generalization later.   Don't see this as an exceptionally critical function to be as general as possible for now.

**Reply by pavlis** [reply]
Another tidbit to preserve here.  Untested, but I am pretty certain I found a clean way to handle this problem  with normalized geometry.   First, note that for the example here l aim to assemble ensembles produced by (1) a source_id and (2) a set of receivers inside a specified circular region (CircularDomain of obspy = $nearSphere of MongodB).  A pseudocode version of the algorithm I think will work is this (uses site collection here but could be channel too):
```
1)  Run $nearSphere query on site for specified lat, lon, and distance cutoff
2)  Create a list of site_id values from the returned cursor.  Call the list "siteids"
3)  Run this query on the wf_Seismogram collection: {'source_id' : source_id, {'site_id' : {'$in' : siteids}}
```
Syntax may not be exactly right, but the point is this exactly the kind of use, I think, the $in operator was designed for.
